devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,237 @@
+        return self._fd
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    def __getattr__(self, name):
+        if name == 'fd' or name == '_fd':
+            raise AttributeError()
+        if name in self.fd:
+            # self.fd[name] = (property, description), calls self.fd[name][0]
+            return self.fd[name][0](self)
+        return self.__getattribute__(name)
+
+    def xreplace(self, rule):
+        if self.is_Function:
+            return super(Differentiable, self).xreplace(rule)
+        else:
+            return self._expr.xreplace(rule)
+
+    def _merge_fd_properties(self, other):
+        merged = {}
+        merged["space_order"] = np.min([getattr(self, 'space_order', 100) or 100,
+                                        getattr(other, 'space_order', 100)])
+        merged["time_order"] = np.min([getattr(self, 'time_order', 100) or 100,
+                                       getattr(other, 'time_order', 100)])
+        merged["indices"] = tuple(filter_ordered(self.indices +
+                                                 getattr(other, 'indices', ())))
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
+        merged["dtype"] = self.dtype
+        return merged
+
+    def __add__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __iadd__(self, other):
+        self._expr = sympy.Add(*[getattr(self, '_expr', self),
+                                 getattr(other, '_expr', other)])
+        return self
+
+    __radd__ = __add__
+
+    def __sub__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          -getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __rsub__(self, other):
+        return Differentiable(sympy.Add(*[-getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __isub__(self, other):
+        self._expr = sympy.Add(*[getattr(self, '_expr', self),
+                                 -getattr(other, '_expr', other)])
+        return self
+
+    def __mul__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __imul__(self, other):
+        self._expr = sympy.Mul(*[getattr(self, '_expr', self),
+                                 getattr(other, '_expr', other)])
+        return self
+
+    __rmul__ = __mul__
```

Comment:
(maybe nitpicking, maybe not): floating point operations are not commutative

Reply:
It's on symbolic  expressions  so should be